build_resolvers/lib/src/analysis_driver_model_uri_resolver.dart

```diff
@@ -0,0 +1,66 @@
+// Copyright (c) 2025, the Dart project authors. Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'package:analyzer/source/file_source.dart';
+// ignore: implementation_imports
+import 'package:analyzer/src/clients/build_resolvers/build_resolvers.dart';
+import 'package:build/build.dart';
+import 'package:path/path.dart' as p;
+
+import 'analysis_driver_model.dart';
+
+const _ignoredSchemes = ['dart', 'dart-ext'];
+
+/// A [UriResolver] on top of [AnalysisDriverModel]'s in-memory filesystem.
+class AnalysisDriverModelUriResolver implements UriResolver {
+  final AnalysisDriverModel analysisDriverModel;
+  AnalysisDriverModelUriResolver(this.analysisDriverModel);
+
+  @override
+  Source? resolveAbsolute(Uri uri, [Uri? actualUri]) {
+    final assetId = parseAsset(uri);
+    if (assetId == null) return null;
+
+    var file = analysisDriverModel.resourceProvider.getFile(assetPath(assetId));
+    return FileSource(file, assetId.uri);
+  }
+
+  @override
+  Uri pathToUri(String path) {
+    var pathSegments = p.posix.split(path);
+    var packageName = pathSegments[1];
+    if (pathSegments[2] == 'lib') {
+      return Uri(
+        scheme: 'package',
+        pathSegments: [packageName].followedBy(pathSegments.skip(3)),
```

Comment:
Again I realize this is just copied from other places, so you don't have to change it here.
I'll just say that we, at least in my opinion, are not doing ourselves any favors by using `followedBy` and `skip`. We're creating a list and two iterables.

Reply:
Thanks.

The analyzer only calls this O(N) times, where N is the number of files; so that won't matter until we've fixed the O(N^2) behaviour. I look forward to getting to that point ;)